### PR TITLE
Supports for`azure_files_identity_based_authentication`, `routing_preference` in `azurerm_storage_account`

### PR DESCRIPTION
**Pull request by yupwei68**
Extend: https://github.com/terraform-providers/terraform-provider-azurerm/pull/8961

Fix: https://github.com/terraform-providers/terraform-provider-azurerm/issues/8960
https://github.com/terraform-providers/terraform-provider-azurerm/issues/9221
https://github.com/terraform-providers/terraform-provider-azurerm/issues/11373

All tests have passed

### CODE REVIEW

File: website/docs/r/storage_account.html.markdown [hunk 1]
```diff
@@ -249,6 +253,40 @@ any combination of `Logging`, `Metrics`, `AzureServices`, or `None`.
 
 ---
 
+A `azure_files_identity_based_authentication` block supports the following:
```
Comment (katbyte) [inline website/docs/r/storage_account.html.markdown:256]: I think this is unclear, is this to configure files authentication? so shouldn't it be
```suggestion
A `azure_files_authentication` block supports the following:
```

File: website/docs/r/storage_account.html.markdown [hunk 2]
```diff
@@ -249,6 +253,40 @@ any combination of `Logging`, `Metrics`, `AzureServices`, or `None`.
...
+
+* `directory_service_options` - (Required) Specifies the directory service used. Possible values are `AADDS` and `AD`.
```
Comment (katbyte) [inline website/docs/r/storage_account.html.markdown:258]:
```suggestion
* `directory_type` - (Required) Specifies the directory service used. Possible values are `AADDS` and `AD`.
```

File: website/docs/r/storage_account.html.markdown [hunk 3]
```diff
@@ -249,6 +253,40 @@ any combination of `Logging`, `Metrics`, `AzureServices`, or `None`.
...
+
+* `active_directory` - (Optional) A `active_directory` block as defined below. Required when `directory_service_options` is `AD`.
+
+---
+
+A `active_directory` block supports the following:
+
+* `azure_storage_sid` - (Required) Specifies the security identifier (SID) for Azure Storage.
```
Comment (katbyte) [inline website/docs/r/storage_account.html.markdown:266]:
```suggestion
* `storage_sid` - (Required) Specifies the security identifier (SID) for Azure Storage.
```

File: website/docs/r/storage_account.html.markdown [hunk 4]
```diff
@@ -249,6 +253,40 @@ any combination of `Logging`, `Metrics`, `AzureServices`, or `None`.
...
+
+* `domain_name` - (Required) Specifies the primary domain that the AD DNS server is authoritative for.
+
+* `domain_sid` - (Required) Specifies the security identifier (SID).
+
+* `domain_guid` - (Required) Specifies the domain GUID.
+
+* `forest_name` - (Required) Specifies the Active Directory forest.
+
+* `net_bios_domain_name` - (Required) Specifies the NetBIOS domain name.
```
Comment (katbyte) [inline website/docs/r/storage_account.html.markdown:276]: netbois is one word
```suggestion
* `netbios_domain_name` - (Required) Specifies the NetBIOS domain name.
```

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks @yupwei68 - left some comments inline

**Review by katbyte — APPROVED** [review]
Thanks @yupwei68 - LGTM 👍

**Comment by singhalss** [comment]
Hi @yupwei68 / @katbyte - While looking at the documentation, I found that we need to specify **storage_sid** as one of the required values in **azure_files_authentication** block. **storage_sid** is generated once the storage account joins AD, not sure what value to specify to storage SID while creating the storage account.

Also, one more doubt, while running terraform in context of service principal (SPN), what permissions are required to be given to SPN to accomplish this task. As per my understanding, service account/SPN will need to have domain join permissions for accomplishing this task and also in automation context, it needs to be executed from a domain join machine.

Please correct me if I am wrong and guide further to make use of this feature in terraform provider.

**Pull request merged**